README.md
# Passpol

[![CircleCI](https://circleci.com/gh/codahale/passpol.svg?style=svg)](https://circleci.com/gh/codahale/passpol)

A Java 12 library for validating passwords against [NIST
SP-800-63B](https://pages.nist.gov/800-63-3/) requirements.

## Add to your project

```xml
<dependency>
  <groupId>com.codahale</groupId>
  <artifactId>passpol</artifactId>
  <version>0.7.0</version>
</dependency>
```

```java
module net.example.yours {
  requires com.codahale.passpol;
}
```

## Use the thing

```java
import com.codahale.passpol.BreachDatabase;
import com.codahale.passpol.PasswordPolicy;

class Example {
  void doIt() {
    final PasswordPolicy policy = new PasswordPolicy(BreachDatabase.haveIBeenPwned(5), 8, 64);
    
    // validate good passwords
    System.out.println(policy.check("this is a good, long password")); 
    
    // validate bad passwords
    System.out.println(policy.check("password"));
    
    // convert a unicode password to a normalized byte array suitable for hashing
    final byte[] bytes = PasswordPolicy.normalize("✊🏻 unicode 🔥 password");
  } 
}
```

## How it works

`PasswordPolicy` checks passwords for minimum and maximum length (i.e. the number of Unicode
codepoints in the string) and can check a set of breach databases to see if the password has been
made public.

The built-in breach databases include an offline list of 100,000 weak passwords from the [SecList
Project](https://github.com/danielmiessler/SecLists)'s collection of breached passwords.

`PasswordPolicy` also provides the means to normalize Unicode passwords into a canonical byte array
representation suitable for inputting into a password hashing algorithm like `bcrypt`.

## License

Copyright © 2017-2019 Coda Hale

Distributed under the Apache License 2.0.


CHANGELOG.md
# Change Log

## 0.7.0: 2019-03-31

* Upgraded to Java 12.
* Rewrote HaveIBeenPwned client to use `java.net.http`.
* Added formal module definition.
* Added support for arbitrary sets of breached passwords.

## v0.6.2: 2018-09-26

* Removed JSR-305 annotations.
* Added automatic module name.

## v0.6.1: 2018-05-01

* Added Have I Been Pwned threshold as a parameter.

## v0.6.0: 2018-04-11

* Moved top-100K password checking to a `BreachDatabase` implementation.
* `PasswordPolicy` now requires a `BreachDatabase` instance.
* Removed `Status.WEAK`.
* Added `BreachDatabase.anyOf` for composing instances.

## v0.5.1: 2018-02-28

* Added a `User-Agent` header to HaveIBeenPwned requests.

## v0.5.0: 2018-02-23

* Changed the HaveIBeenPwned breach database to use their new, _k-anonymous_ backend.
* Improve memory usage and performance slightly.

## v0.4.0: 2017-12-26

* Changed `PasswordPolicy` to return the reason for rejecting a password, if any.
* Made `PasswordPolicy::normalize` static.
* Added support for using breach databases, including Have I Been Pwned?.

## v0.3.0: 2017-11-07

* Increased weak password list to 100,000.

## v0.2.1: 2017-10-28

* Simplified code point counting.

## v0.2.0: 2017-06-26

* Remove Guice dependency.

## v0.1.4: 2017-04-21

* Add optional JSR305 annotations.
* Fix resource path of weak passwords list.

## v0.1.3: 2017-04-20

* Improve password loading.

## v0.1.2: 2017-04-19

* Improve codepoint counting.

## v0.1.1: 2017-04-19

* Count Unicode codepoints, not characters.

## v0.1.0: 2017-04-19

* Initial release.

src/test/java/com/codahale/passpol/Download.java
/*
 * Copyright © 2018 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.passpol;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Paths;

public class Download {

  private static final String URL =
      "https://raw.githubusercontent.com/danielmiessler/SecLists/master/Passwords/Common-Credentials/10-million-password-list-top-100000.txt";
  private static final String FILE = "src/main/resources/com/codahale/passpol/weak-passwords.txt";

  public static void main(String[] args) throws IOException, InterruptedException {
    final HttpRequest req = HttpRequest.newBuilder().uri(URI.create(URL)).GET().build();
    HttpClient.newHttpClient().send(req, HttpResponse.BodyHandlers.ofFile(Paths.get(FILE)));
  }
}


src/test/java/com/codahale/passpol/tests/StatusTest.java
package com.codahale.passpol.tests;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.codahale.passpol.Status;
import org.junit.jupiter.api.Test;

class StatusTest {

  @Test
  void onlyOKisAcceptable() {
    assertTrue(Status.OK.isAcceptable());
    assertFalse(Status.TOO_LONG.isAcceptable());
    assertFalse(Status.TOO_SHORT.isAcceptable());
    assertFalse(Status.WEAK.isAcceptable());
    assertFalse(Status.BREACHED.isAcceptable());
  }
}


src/test/java/com/codahale/passpol/benchmarks/Benchmarks.java
/*
 * Copyright © 2018 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.passpol.benchmarks;

import com.codahale.passpol.PasswordPolicy;
import com.codahale.passpol.Status;
import java.util.concurrent.TimeUnit;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
public class Benchmarks {

  private static final String PASSWORD = "this is a decent password";

  private final PasswordPolicy policy = new PasswordPolicy();

  @Benchmark
  public long charsStream() {
    return PASSWORD.codePoints().count();
  }

  @Benchmark
  public long charsRange() {
    return PASSWORD.codePointCount(0, PASSWORD.length());
  }

  @Benchmark
  public Status check() {
    return policy.check(PASSWORD);
  }

  @Benchmark
  public byte[] normalize() {
    return PasswordPolicy.normalize(PASSWORD);
  }
}


src/main/java/com/codahale/passpol/Status.java
/*
 * Copyright © 2017 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.passpol;

/** The status of a given candidate password. */
public enum Status {
  /** The candidate password is acceptable. */
  OK,

  /** The candidate password is too short. */
  TOO_SHORT,

  /** The candidate password is too long. */
  TOO_LONG,

  /** The candidate password is on the weak passwords list. */
  WEAK,

  /** The candidate password has previously appeared in a data breach. */
  BREACHED
}


src/main/java/com/codahale/passpol/Top100K.java
/*
 * Copyright © 2018 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.passpol;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.Set;
import java.util.stream.Collectors;

class Top100K implements BreachDatabase {
  private static final Top100K INSTANCE = new Top100K();
  private final Set<String> passwords;

  static Top100K instance() {
    return INSTANCE;
  }

  private Top100K() {
    try (var in = getClass().getResourceAsStream("weak-passwords.txt");
        var r = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8))) {
      this.passwords = r.lines().collect(Collectors.toUnmodifiableSet());
    } catch (IOException e) {
      throw new IllegalStateException(e);
    }
  }

  @Override
  public boolean contains(String password) {
    return passwords.contains(password);
  }
}


src/main/java/com/codahale/passpol/PasswordSet.java
/*
 * Copyright © 2018 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.passpol;

import java.text.Normalizer;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class PasswordSet implements BreachDatabase {

  private final Set<String> passwords;

  PasswordSet(Stream<String> passwords) {
    this.passwords = passwords.map(PasswordSet::normalize).collect(Collectors.toUnmodifiableSet());
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    var that = (PasswordSet) o;
    return Objects.equals(passwords, that.passwords);
  }

  @Override
  public int hashCode() {
    return Objects.hash(passwords);
  }

  @Override
  public String toString() {
    return passwords.toString();
  }

  @Override
  public boolean contains(String password) {
    return passwords.contains(normalize(password));
  }

  static String normalize(String s) {
    return Normalizer.normalize(s, Normalizer.Form.NFKC);
  }
}


benchmarks/src/main/java/com/codahale/Benchmarks.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.codahale;

import com.codahale.passpol.PasswordPolicy;
import java.io.IOException;
import java.util.concurrent.TimeUnit;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
public class Benchmarks {

  private static final String PASSWORD = "this is a decent password";
  private final PasswordPolicy policy;

  public Benchmarks() {
    try {
      this.policy = new PasswordPolicy();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  @Benchmark
  public boolean test() {
    return policy.test(PASSWORD);
  }

  @Benchmark
  public byte[] normalize() {
    return policy.normalize(PASSWORD);
  }
}


src/main/java/com/codahale/passpol/BreachDatabase.java
/*
 * Copyright © 2018 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.passpol;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.http.HttpClient;
import java.nio.charset.StandardCharsets;
import java.util.Collection;

/** A database of passwords found in data breaches. */
public interface BreachDatabase {

  /**
   * Returns whether or not the database contains the given password.
   *
   * @param password a candidate password
   * @return {@code true} if the database contains {@code password}
   * @throws IOException if there was a problem communicating with the database
   */
  boolean contains(String password) throws IOException;

  /**
   * A client for <a href="https://haveibeenpwned.com/">Have I Been Pwned</a>'s online password
   * checking. Uses a k-anonymous API which transmits only 20 bits of a password hash.
   *
   * @return an online database of breached passwords
   */
  static BreachDatabase haveIBeenPwned() {
    return haveIBeenPwned(1);
  }

  /**
   * A client for <a href="https://haveibeenpwned.com/">Have I Been Pwned</a>'s online password
   * checking. Uses a k-anonymous API which transmits only 20 bits of a password hash.
   *
   * @param threshold The number of breaches a password can be found in which makes it invalid.
   * @return an online database of breached passwords
   */
  static BreachDatabase haveIBeenPwned(int threshold) {
    return haveIBeenPwned(HttpClient.newHttpClient(), threshold);
  }

  /**
   * A client for <a href="https://haveibeenpwned.com/">Have I Been Pwned</a>'s online password
   * checking. Uses a k-anonymous API which transmits only 20 bits of a password hash.
   *
   * @param client The HTTP client to use
   * @param threshold The number of breaches a password can be found in which makes it invalid.
   * @return an online database of breached passwords
   */
  static BreachDatabase haveIBeenPwned(HttpClient client, int threshold) {
    return new HaveIBeenPwned(client, threshold);
  }

  /**
   * Returns an offline database of the given passwords.
   *
   * @param passwords a collection of unusable passwords
   * @return an offline database of the given passwords
   */
  static BreachDatabase passwordSet(Collection<String> passwords) {
    return new PasswordSet(passwords.stream());
  }

  /**
   * Returns an offline database of the 100,000 most common passwords.
   *
   * @return an offline database of the 100,000 most common passwords
   */
  static BreachDatabase top100K() {
    try (var in = BreachDatabase.class.getResourceAsStream("weak-passwords.txt");
        var r = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8))) {
      return new PasswordSet(r.lines());
    } catch (IOException e) {
      throw new IllegalStateException(e);
    }
  }

  /**
   * Returns a database which checks the given databases in order.
   *
   * @param databases a set of databases
   * @return a database which checks the given databases in order
   */
  static BreachDatabase anyOf(BreachDatabase... databases) {
    return password -> {
      for (var database : databases) {
        if (database.contains(password)) {
          return true;
        }
      }
      return false;
    };
  }
}


src/main/java/com/codahale/passpol/HaveIBeenPwned.java
/*
 * Copyright © 2018 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.passpol;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse.BodyHandlers;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class HaveIBeenPwned implements BreachDatabase {
  private static final URI BASE_URI = URI.create("https://api.pwnedpasswords.com/range/");
  private final HttpClient client;
  private final int threshold;

  HaveIBeenPwned(HttpClient client, int threshold) {
    this.client = client;
    this.threshold = threshold;
  }

  @Override
  public boolean contains(String password) throws IOException {
    try {
      var hash = hex(MessageDigest.getInstance("SHA1").digest(PasswordPolicy.normalize(password)));
      var pattern = Pattern.compile("^" + hash.substring(5) + ":([\\d]+)$");
      var request =
          HttpRequest.newBuilder()
              .GET()
              .uri(BASE_URI.resolve(hash.substring(0, 5)))
              .header("User-Agent", "passpol")
              .build();
      var response = client.send(request, BodyHandlers.ofLines());
      if (response.statusCode() != 200) {
        throw new IOException("Unexpected response from server: " + response.statusCode());
      }
      return response
          .body()
          .map(pattern::matcher)
          .flatMap(Matcher::results)
          .map(r -> r.group(1))
          .mapToInt(Integer::parseInt)
          .anyMatch(t -> t >= threshold);
    } catch (NoSuchAlgorithmException | InterruptedException e) {
      throw new IOException(e);
    }
  }

  private static final char[] HEX = {
    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'
  };

  private static String hex(byte[] bytes) {
    var b = new StringBuilder(bytes.length * 2);
    for (var v : bytes) {
      b.append(HEX[(v & 0xFF) >> 4]);
      b.append(HEX[(v & 0x0F)]);
    }
    return b.toString();
  }
}


src/main/java/com/codahale/passpol/PasswordPolicy.java
/*
 * Copyright © 2018 Coda Hale (coda.hale@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codahale.passpol;

import java.io.IOException;
import java.nio.charset.StandardCharsets;

/**
 * A password policy which validates candidate passwords according to NIST's draft {@code
 * SP-800-63B}, which recommend passwords have a minimum required length, a maximum required length,
 * ad be checked against a list of weak passwords ({@code SP-800-63B 5.1.1.2}).
 *
 * <p>This uses a static list of 10,000 weak passwords downloaded from Carey Li's NBP project.
 *
 * @see <a href="https://pages.nist.gov/800-63-3/">Draft NIST SP-800-63B</a>
 * @see <a href="https://cry.github.io/nbp/">NBP</a>
 */
public class PasswordPolicy {

  /** The recommended minimum password length, per {@code SP-800-63B 5.1.1.2}. */
  public static final int RECOMMENDED_MIN_LENGTH = 8;

  /** The recommended maximum password length, per {@code SP-800-63B 5.1.1.2}. */
  public static final int RECOMMENDED_MAX_LENGTH = 64;

  private final int minLength;
  private final int maxLength;
  private final BreachDatabase breachDatabase;

  /**
   * Creates a {@link PasswordPolicy} with a minimum password length of {@code 8} and a maximum
   * password length of {@code 64}, as recommended in {@code SP-800-63B 5.1.1.2}. Uses the offline
   * database of weak passwords.
   *
   * @see BreachDatabase#top100K()
   * @see PasswordPolicy#RECOMMENDED_MIN_LENGTH
   * @see PasswordPolicy#RECOMMENDED_MAX_LENGTH
   */
  public PasswordPolicy() {
    this(BreachDatabase.top100K(), RECOMMENDED_MIN_LENGTH, RECOMMENDED_MAX_LENGTH);
  }

  /**
   * Creates a {@link PasswordPolicy} with the given password length requirements.
   *
   * @param minLength the minimum length of passwords
   * @param maxLength the maximum length of passwords
   * @param breachDatabase a {@link BreachDatabase} instance
   */
  public PasswordPolicy(BreachDatabase breachDatabase, int minLength, int maxLength) {
    if (maxLength < minLength) {
      throw new IllegalArgumentException("minLength must be less than maxLength");
    }
    this.breachDatabase = breachDatabase;
    this.minLength = minLength;
    this.maxLength = maxLength;
  }

  /**
   * Normalizes the given password as Unicode NFKC and returns it as UTF-8 encoded bytes, ready to
   * be passed to a password hashing algorithm like {@code bcrypt}.
   *
   * <p>This is the process recommended in {@code NIST SP-800-63B 5.1.1.2}.
   *
   * @param password an arbitrary string
   * @return a series of bytes suitable for hashing
   */
  public static byte[] normalize(String password) {
    return PasswordSet.normalize(password).getBytes(StandardCharsets.UTF_8);
  }

  /**
   * Checks the acceptability of a candidate password.
   *
   * @param password a candidate password
   * @return the status of {@code password}
   */
  public Status check(String password) {
    var len = password.codePointCount(0, password.length());

    if (len < minLength) {
      return Status.TOO_SHORT;
    }

    if (len > maxLength) {
      return Status.TOO_LONG;
    }

    try {
      if (breachDatabase.contains(password)) {
        return Status.BREACHED;
      }
    } catch (IOException e) {
      return Status.OK;
    }

    return Status.OK;
  }
}
